File: ai/app/controllers/itinerary_controller.py
from sqlalchemy.orm import Session
from app.schemas.itinerary import (
    InteractionEmbeddingRequest, InteractionEmbeddingResponse,
    ItineraryGenerateRequest, ItineraryGenerateResponse,
    MishapRecoverRequest, MishapRecoverResponse,
)
from app.services.interaction_service import create_interaction_embedding
from app.services import itinerary_service

"""
Itinerary & Interaction Controllers — WeTravel AI Backend
Layer: Controller (Routes → Controller → Service)
"""


def handle_interaction_embedding(
    payload: InteractionEmbeddingRequest,
    db: Session,
) -> InteractionEmbeddingResponse:
    result = create_interaction_embedding(
        db=db,
        group_id=payload.group_id,
        source_type=payload.source_type,
        source_id=payload.source_id,
        summary_text=payload.summary_text,
    )
    return InteractionEmbeddingResponse(
        status="success",
        message="Interaction embedding generated and saved.",
        group_id=result["group_id"],
        source_type=result["source_type"],
        vector_dimensions=result["vector_dimensions"],
    )


async def handle_generate_itinerary(
    payload: ItineraryGenerateRequest,
    db: Session,
) -> ItineraryGenerateResponse:
    itineraries = await itinerary_service.generate_itineraries(
        group_id=payload.group_id,
        origin=payload.origin,
        destination=payload.destination,
        start_date=payload.start_date,
        end_date=payload.end_date,
        member_count=payload.member_count,
        group_stats=payload.group_stats or {},
        constraints=payload.constraints or {},
    )
    return ItineraryGenerateResponse(
        group_id=payload.group_id,
        destination=payload.destination,
        itineraries=itineraries,
        total_variants=len(itineraries),
    )


async def handle_mishap_recovery(
    payload: MishapRecoverRequest,
    db: Session,
) -> MishapRecoverResponse:
    result = await itinerary_service.recover_itinerary(
        group_id=payload.group_id,
        itinerary=payload.itinerary,
        missed_item_id=payload.missed_item_id,
        current_time=payload.current_time,
    )
    return MishapRecoverResponse(
        group_id=payload.group_id,
        itinerary=result,
        rescheduled_count=result["constraints"].get("rescheduled_count", 0),
        message=result["summary"],
    )

File: backend/src/controllers/itinerary.controller.js
const itineraryService = require('../services/itinerary.service');

/**
 * Itinerary Controller — WeTravel Backend
 * Layer: Controller
 */

/** POST /api/trips/:tripId/itinerary/generate */
const generateItineraries = async (req, res, next) => {
  try {
    const { origin, destination, startDate, endDate, memberCount, constraints } = req.body;
    const itineraries = await itineraryService.generateItineraries(req.user.id, req.params.tripId, {
      origin, destination, startDate, endDate, memberCount, constraints,
    });
    res.status(201).json({
      status: 'success',
      message: `${itineraries.length} itinerary variants generated successfully.`,
      data: { itineraries },
    });
  } catch (err) { next(err); }
};

/** GET /api/trips/:tripId/itineraries */
const getItineraries = async (req, res, next) => {
  try {
    const itineraries = await itineraryService.getItineraries(req.user.id, req.params.tripId);
    res.status(200).json({ status: 'success', data: { itineraries } });
  } catch (err) { next(err); }
};

/** GET /api/trips/:tripId/itineraries/:id */
const getItinerary = async (req, res, next) => {
  try {
    const itinerary = await itineraryService.getItineraryById(req.user.id, req.params.tripId, req.params.id);
    res.status(200).json({ status: 'success', data: { itinerary } });
  } catch (err) { next(err); }
};

/** PATCH /api/trips/:tripId/itineraries/:id/select */
const selectItinerary = async (req, res, next) => {
  try {
    const itinerary = await itineraryService.selectItinerary(req.user.id, req.params.tripId, req.params.id);
    res.status(200).json({ status: 'success', message: 'Itinerary selected.', data: { itinerary } });
  } catch (err) { next(err); }
};

/** POST /api/trips/:tripId/itineraries/:id/suggest */
const suggestChanges = async (req, res, next) => {
  try {
    const { origin, destination, startDate, endDate, memberCount, constraints } = req.body;
    const itineraries = await itineraryService.suggestChanges(req.user.id, req.params.tripId, req.params.id, {
      origin, destination, startDate, endDate, memberCount, constraints,
    });
    res.status(201).json({ status: 'success', message: 'New itineraries generated with your changes.', data: { itineraries } });
  } catch (err) { next(err); }
};

/** POST /api/trips/:tripId/itineraries/mishap */
const recoverFromMishap = async (req, res, next) => {
  try {
    const { itineraryId, missedItemId, currentTime } = req.body;
    if (!itineraryId || !missedItemId || !currentTime) {
      return res.status(400).json({ status: 'error', message: 'itineraryId, missedItemId, and currentTime are required.' });
    }
    const recovery = await itineraryService.recoverFromMishap(req.user.id, req.params.tripId, {
      itineraryId, missedItemId, currentTime,
    });
    res.status(201).json({ status: 'success', message: 'Recovery itinerary generated.', data: { itinerary: recovery } });
  } catch (err) { next(err); }
};

module.exports = { generateItineraries, getItineraries, getItinerary, selectItinerary, suggestChanges, recoverFromMishap };
